Haddock recompilation avoidance

FinleyMcIlwaine opened the pull request:
Haddock no longer writes compilation files by default, so we do not need to pass tmp dirs for `-hidir`, `-stubdir`, and `-odir` via `--optghc`. Indeed, we do not *want* to do so, since it results in recompilation for every invocation of Haddock via Cabal. This commit stops this from happening for Haddock versions 2.28 or greater (when Hi Haddock was introduced).

This commit also prevents the definition of the `__HADDOCK_VERSION__` macro when invoking GHC through Haddock by default, which necessary to avoid recompilation during documentation generation, since the macro invalidates existing build results. Since a limited set of users may rely on this macro still, we introduce the `--haddock-version-cpp` flag and `haddock-version-cpp:` cabal.project field, which enable the definition of the `__HADDOCK_VERSION__` macro when invoking GHC through Haddock.

This commit also renames the `--haddock-lib` flag to `--haddock-resources-dir` (and `haddock-lib:` cabal.project field to `haddock-resources-dir:`), and adds this flag to the users guide since it was missing an entry. This also allows us to add this field to `cabal-install:test:integration-tests2`, since it is no longer ambiguous with the `--lib` flag.

This commit also causes `documentation: true` or `--enable-documentation` to imply `-haddock` for GHC.

Also, since Haddock >= 2.29 is renaming `--lib` to `--resources-dir`, this commit switches the flag provided to Haddock using a backwards compatible condition based on the Haddock version.

---
## QA Notes

Consider a Haskell package with dependencies. We should be able to simply configure the package so that doing `cabal build` followed by `cabal haddock` does not trigger a recompilation. Use the following `deps.cabal` file:
```cabal
cabal-version: 3.6
name: deps
version: 0.1.0

library
  build-depends:
      base
    , acme-missiles
  exposed-modules: Lib
```

The following `cabal.project` file:
```cabal
packages: .

package *
  documentation: True
  haddock-options: --optghc=-ddump-hi-diffs
```

And any `Lib.hs` will do:
```haskell
-- | Lib reexports 'launchMissiles'
module Lib (myMissiles, launchMissiles) where

import Acme.Missiles

-- | Another 'launchMissiles'
myMissiles :: IO ()
myMissiles = launchMissiles
```

This configuration should avoid recompilation during documentation generation. To verify this, with a package such as above, just run `cabal build`. Since documentation is enabled, the package will first build then generate Haddocks. Since Haddock is passing `-ddump-hi-diffs` to GHC, you should see output indicating that no recompilation checks have triggered during the Haddock phase:
```
Module flags unchanged
Optimisation flags changed; ignoring
HPC flags unchanged
signatures to merge in unchanged
[]
implementing module unchanged
Checking interface for module Prelude base
Module fingerprint unchanged
Checking interface for module Acme.Missiles cm-mssls-0.3-9160cec8
Module fingerprint unchanged
Checking interface for module GHC.Types ghc-prim
Module fingerprint unchanged
```

This means recompilation was successfully avoided.

**NOTE:** I am working on making `no-tmp-comp-dir` the default behavior for Haddock 2.29, but until then it is necessary to avoid recompilation.

---

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [x] The documentation has been updated, if necessary.
* [x] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included.

Fixes  #9175

dcoutts commented:
Is it guaranteed that haddock will _not_ write/overwrite any compilation files? Its ok for haddock to read files from the normal compilation output, but never ok to write them.

So how does this work? Does the `cabal haddock` command now run the compilation first (with its usual compilation avoidance)?

> Haddock no longer writes compilation files by default

I'd feel happier if haddock had a flag that promises to never write compilation files, and instead fail if it would need to do so. This would give us a "belt & braces" solution, in case something gets messed up in future. Otherwise we could end up with some hard-to-debug problems.

FinleyMcIlwaine commented:
> Is it guaranteed that haddock will _not_ write/overwrite any compilation files?

I do think this is guaranteed (for Haddock versions >= 2.28.0, which is when "Hi Haddock" was introduced). Haddock always uses GHC's `-fno-code` backend (unless the user explicitly enables a different GHC backend), which will ["omit code generation (and all later phases) altogether"](https://downloads.haskell.org/ghc/9.6.2/docs/users_guide/phases.html#ghc-flag--fno-code). It *does* generate code for dependencies of modules containing Template Haskell splices, but as far as I can tell those compilation results are put in a temporary directory by GHC and do not overwrite anything.

> Does the cabal haddock command now run the compilation first (with its usual compilation avoidance)?

This PR doesn't add or remove any calls to tools for the cabal haddock command. In that regard, the cabal haddock command still works just as it did before, meaning it will

1. make sure installed dependency packages are up to date (in the same way the cabal build command does), and
2. run haddock on the local package.

The important bit is that with "Hi Haddock", Haddock generates documentation from module interfaces that it gets from the GHC API. The GHC API will do the normal compilation avoidance things while retrieving those interfaces.

The problem that this PR addresses is that cabal was passing temporary `-hidir` directories to GHC through Haddock, meaning Haddock's calls to GHC would always result in recompilation. This patch just stops those temporary directories from being passed.

As an aside: Haddock has been running its GHC session in a temporary directory by default since version 2.23.0, shipped with GHC 8.8, until version 2.29 (pending [this change](https://gitlab.haskell.org/ghc/haddock/-/merge_requests/38)). So, as long as this cabal patch is not used in combination with a GHC 8.6.5 bundled Haddock (version 2.22 or lower), this change should be backwards compatible and avoid overwriting any existing compilation files.

> I'd feel happier if haddock had a flag that promises to never write compilation files, and instead fail if it would need to do so.

Is the `-fno-code` usage enough of a guarantee here? As long as `-fno-code` does not write compilation files, then Haddock won't either.

michaelpj commented:
It would be really nice if oblivious users could benefit from this change. As far as I can tell, the main reason why that won't be true is `haddock-no-version-cpp`. In order for user to benefit by default, we'd have to set that option to `false` by default. I realise that's a breaking change, but if we don't do this I predict that only a small fraction of users will actually benefit from this change, which seems a shame given that the flag is apparently very rarely necessary.

Personally, I would be willing to even just deprecate `__HADDOCK_VERSION__` if it means we get reliable recompilation avoidance for all users. 19 Hackage packages! Maybe we can just purge it from them.

(I took a quick look at some of the search results. It seems like most of the usage is "hide this bit of documentation from Haddock because Haddock has a bug". It seems to me that the version of GHC could often be used as a proxy for the problematic Haddock version in those cases.)

phadej commented:
> This PR doesn't add or remove any calls to tools for the cabal haddock command. 

And

> so we do not need to pass tmp dirs for -hidir, -stubdir, and -odir via --optghc. 

are slightly contradictory. You don't remove any calls, but you do change them, as far as I can see, unconditionally.

For old GHCs & haddock combinations I would expect *no changes in behaviour*. New `Cabal` library versions are used 
to work with old GHCs & haddocks.

FinleyMcIlwaine commented:
> Personally, I would be willing to even just deprecate `__HADDOCK_VERSION__` if it means we get reliable recompilation avoidance for all users.

What would the deprecation strategy look like for this? Could I just change `haddock-no-version-cpp` to `haddock-version-cpp` and make it default to `false`?

FinleyMcIlwaine commented:
> I don't see any version checks

I will add a version check that only prevents the tmp dirs being passed for Haddock >= 2.29.

andreabedini commented:
> What would the deprecation strategy look like for this? Could I just change `haddock-no-version-cpp` to `haddock-version-cpp` and make it default to `false`?

Putting on my cowboy hat, I'd say I am ok with this, this is one of the behaviours I want to see changing.

Setting `haddock-version-cpp: true` gives users a very easy way out. They had the resources to update to the latest version of cabal, they will have the resouces to add that line to cabal.project :relaxed:  :cowboy_hat_face:

FinleyMcIlwaine commented:
> Putting on my cowboy hat, I'd say I am ok with this

I agree 🤠

I've changed this to avoid defining the `__HADDOCK_VERSION__` macro by default and introduced the `--haddock-version-cpp` flag and `haddock-version-cpp:` cabal.project field, defaulting to `False`. I updated the PR description QA notes and changelog entry to reflect this change.

phadej commented:
I kindly ask you to think a bit.  `__HADDOCK_VERSION__` was there mostly to indicate that the compiler is "haddock". No-one actually checks the actual version. 

As far as I can tell the functionality has **only** been used to circumvent bugs in haddock: https://hackage-search.serokell.io/?q=__HADDOCK_VERSION__

`base`:

```haskell
#if !defined(__HADDOCK_VERSION__)
    -- workaround https://github.com/haskell/haddock/issues/680
```

`generics-sop`:

```haskell
#if !(defined(__HADDOCK_VERSION__)) || MIN_TOOL_VERSION_haddock(2,14,0)
  , Proxy(..) -- hidden from old Haddock versions, because it triggers an internal error
#endif
```

(Note how the tool version is checked, __HADDOCK_VERSION__ is used for whether haddock runs. `MIN_TOOL_VERSION_haddock` macro is always defined! so the version in __HADDOCK_VERSION__ is actually now redundant.

`streamly`:

```haskell
#ifdef __HADDOCK_VERSION__
#undef INSPECTION
#endif
#ifdef INSPECTION
{-# LANGUAGE TemplateHaskell #-}
{-# OPTIONS_GHC -fplugin Test.Inspection.Plugin #-}
#endif
```

This case is also important to think through: here it's essential to know whether package is compiled by haddock, as it simply won't work. So having haddock compile the source and not setting `__HADDOCK_VERSION__` will break this.

---

Either the way haddock-uses-ghc-lib is buggy, or how haddock processed the result data is buggy. The latter is quite likely to exist also in the future and people may want to change the code for haddock to succeed also with hi, haddock. Do they?

In that light, I don't really see a real need for any configuration. You can be conservative and have `haddock-recompile` which would **always** recompile for haddock (not "this will likely trigger recompilation during documentation generation.", but **always**).

There *could* be a way to distinguish whether the source is processed by haddock, but `haddock-version-cpp` is a bit indirect. Also it's documentation says:

> This will likely trigger recompilation during documentation generation.

where for bug circumention it should be assertive as:

> This will trigger recompilation during documentation generation.

I'd like `haddock-recompiles` or `haddock-separate-compilation` like of clear cut option.

And the `haddock-recompiles: False` case (when supported) should in fact tell `haddock+ghc` to not try to compile anything. That's the flag @dcoutts asks for. I agree, you should be explicit about that.

---

There is few examples where people are lazy and add additional imports with `__HADDOCK_VERSION__` to make haddock link stuff in docs, but that is just lazyness, they should use qualified names: a bit ugly, but correct way of doing that.

michaelpj commented:
> haddock-version-cpp: cabal.project field, defaulting to False

I note that this can't be _project_-level configuration. Packages that rely on this may exist on Hackage - how are you supposed to know how to haddock them? How is Hackage supposed to know how to haddock them? If we were going to make it configurable, it seems to me that it would need to be a `.cabal` file field.

> This case is also important to think through: here it's essential to know whether package is compiled by haddock, as it simply won't work.

This case is indeed interesting, because it's about compiler plugins. Perhaps I'm being naive, but I would _hope_ that the upshot of HI Haddock is that most bugs such as this go away: haddock doesn't need to know how to handle compiler plugins, because GHC has already dealt with it. 

So hopefully there will be less need to work around Haddock in the future. But also suppose that we made it impossible to ever detect if Haddock is running. That would lose us the ability to work around some Haddock bugs... which would mean that in some edge cases (19! in all of Hackage!) some things might not be able to be documented. Perhaps that's an acceptable cost.

phadej commented:
> 19! in all of Hackage!)

It's 19 occurences of `__HADDOCK_VERSION__`.

- `Cabal`: matches the code to set the flag. false positive.
- `Cabal-ide-backend`: ditto.
- `base`: working around bug https://github.com/haskell/haddock/issues/680 which is closed, so probably can be removed (if not already_
- `concurrent-extra`: imports for links. invalid. (instead of `'mask'` use `'Control.Exception.mask'` in haddock comments)
- `contrainers`: imports for links
- `dimensions`: unclear
- `dlist`: imports
- `easytensor`: unclear (same author as `dimensions`)
- `generics-sop`: bug with older haddock
- `glambda`: imports
- `hackport`: `Cabal` copy
- `haskell-gi`: unclear (code generation)
- `ngx-export-log`: see below. not an issue with hi haddock
- `paths`: import
- `sop-core`: bug with older haddock
- `streamly`: not an issue with hi haddock. detection of "ghci" like mode.
- `strict-containers`: imports
- `unbound-delays`: imports
- `usb`: imports. This one is tricky, as the imports come from potentially different modules.

So:
- 8 of 19 is imports (almost half!)
- 3 is Cabal copies
- 3 *fixed* bugs
- 2 ghci-like mode detections
- 3 other!?
  - two (`dimensions` and `easytensor`) of which are the same (something weird happening)
  - and `haskell-gi` is code generation, I suspect that `ENABLE_OVERLOADING` is disabled to actually change documentation to not be littered with overloads. I'm not sure that it's great idea though.

I'd say the cost is very low.

`haddock` has been buggy, but I don't know how much my (past) bad experience with it 
is due it trying to be GHC versus the other bugs.

---

`ngc-export-log`:

```
-- Some tools such as hls, haddock, and ghci run interactive linking against C
-- functions plugin_ngx_http_haskell_log() and plugin_ngx_http_haskell_log_r()
-- when loading Log.hs. In Log.hs, TH declarations from Log/Gen.hs which make
-- calls to those C functions, get instantiated. Obviously, linking fails as
-- soon as we don't have a library to expose the functions because such a
-- library is built by Nginx and we don't want to use Nginx at this step.
--
-- In this workaround, the C functions get replaced by stubs when running by
-- hls or haddock. This prevents interactive linking in Log.hs. It's easy to
-- detect that the code is being run by hls or haddock: the tools define their
-- own C macro declarations __GHCIDE__ and __HADDOCK_VERSION__ respectively. To
-- prevent interactive linking in ghci, pass one of the two macro declarations
-- in an appropriate option, e.g.
--
-- cabal repl --ghc-options=-D__GHCIDE__ --repl-options=-fobject-code
```

domenkozar commented:
Hitting memory limits due to haddock, this can't come soon enough.

Kleidukos commented:
@FinleyMcIlwaine Where are we standing with this PR? Anything you need?

geekosaur commented on changelog.d/pr-9177:
```diff
@@ -0,0 +1,28 @@
+  of Haddock via Cabal. We now stop this from happening for Haddock versions
+  2.28 and greater, since that is when Hi Haddock was introduced.
+
+* We no longer defined the `__HADDOCK_VERSION__` macro when invoking GHC through
```
```suggestion
* We no longer define the `__HADDOCK_VERSION__` macro when invoking GHC through
```

alt-romes commented:
Here's the status of this PR, discussing the things which aren't yet addressed.
Please correct me if I'm misinterpreting the discussion.

With this patch we no longer use temporary directories for compilation artifacts in the haddock invocations.
Instead, haddock should reuse the interface files produced by the GHC invocation.

However, if haddock invocations include `-D__HADDOCK_VERSION__` in addition to the
flags used by GHC, recompilation will be triggered because the preprocessor flags changed.

It turns out that barely any packages rely on `__HADDOCK_VERSION__`, but it is
nonetheless an important mechanism to workaround haddock bugs.

Therefore,

1. We want to introduce a *per-package* flag `--haddock-version-cpp`.
Packages should be able to say "I need `__HADDOCK_VERSION__` for haddock".

2. If they do, Cabal invocations of haddock for that package should include
`-D__HADDOCK_VERSION__` (despite that triggering recompilation).

I think `haddock-recompilation` is not what we want. Recompilation is a side
effect of needing HADDOCK_VERSION, but it's not something we want to happen in
its own right.

By default, packages don't run into haddock bugs and therefore needn't use
`__HADDOCK_VERSION__`, and thus those haddock for those packages can reuse
existing compilation artifacts rather than recompile them.

Ultimately, we'd also like that:

3. (Newer) Haddock versions to have a flag like `-fnever-overwrite` to ensure
    haddock will never overwrite GHC produced objects for Cabal to use. This
    would serve as a sanity check that our invocations of Haddock don't mess
    things up.

4. `--no-tmp-comp-dir` be used by default if Haddock supports it (>=2.29?)

The points which I've bulleted are the ones I believe I need to act on in order
to finish this PR. (2) is already done, and (1) must be made a package option
rather instead of project option. (3) needs a change in haddock and in Cabal,
and (4) in Cabal only.

I'm not sure if (3) is what @phadej and @dcoutts really meant on.

michaelpj commented:
I'm still wondering if we can get away with being more aggressive about `__HADDOCK_VERSION__`. From my reading of @phadej 's survey it's not clear that there are _any_ actually necessary uses. Plus, if I understand rightly, this macro is the only thing keeping the whole code path where haddock compiles modules alive! We could just delete all of that code! Certainly, I doubt we would introduce this feature today given the cost.

Perhaps a test is in order: we could compile the packages that @phadej lists without `__HADDOCK_VERSION__` and check if they all look reasonable.

alt-romes commented:
I have to say I'm quite swayed by the idea of deleting the code path, though not sure to what extent it can really all go.

> Perhaps a test is in order: we could compile the packages that @phadej lists without `__HADDOCK_VERSION__` and check if they all look reasonable.

I'll do this, but to what extent will I be able to observe the problems worked around without wasting too much time? It appears all the ones referred to by id are fixed. And perhaps I can go from Oleg's analysis and just look at those that are unclear what are doing with `__HADDOCK_VERSION__`.

michaelpj commented:
Perhaps I'm wrong, but AIUI since haddock is tied to a particular GHC release, there's no need to support old GHCs. The only other reason I can see for needing the old code path is if GHC has not written the information into the interface files, but we could just fail in that case (and this PR then ensures that Cabal will always pass the right flags so that that doesn't happen).

alt-romes commented:
As Oleg noted, `haskell-gi` uses `HADDOCK_VERSION` to disable overloading when building haddocks:

```
commit 64601c4e5a63ef29c779f560a16cfd4a1400a7c5
Author: Iñaki García Etxebarria <garetxe@gmail.com>
Date:   Sat Aug 12 11:24:05 2017 +0200

    Build haddocks without overloading

    This is both faster, and makes the generated documentation less confusing.
```

I tested `haskell-gi` without `HADDOCK_VERSION` on `gi-vte` and the documentation looks reasonable, but is in fact more crowded because of many overloaded label definitions.

I think it isn't that bad, and in the time for Cabal to release a version that does away with `HADDOCK_VERSION`, `haddock` could grow a mode that hides overloaded labels (if agreed, I could implement).

The remaining two "problematic" packages, `easytensor` and `dimensions`, use a plugin called `constraint-deriving`. For all cases except one, `HADDOCK_VERSION` is to provide instances with undefined methods manually for when haddock is run -- this becomes obsolete now with a haddock which works in the presence of plugins too by looking at interfaces only.

However, there is exactly one case where the author leverages these "dummy instances" (only used when `HADDOCK_VERSION` is defined) to add custom haddocks to these instances:

```haskell
#if defined(__HADDOCK__) || defined(__HADDOCK_VERSION__)

{- |
Although @Enum (Idx d)@ requires @BoundedDim d@, it does not use @maxBound@
when @(d ~ XN m)@.
You can use list comprehensions safely for known dims
(@(k ~ Nat)@ or @(d ~ N d)@),
but you may get an index larger than your struct to be indexed when @d ~ XN m@.
 -}
deriving instance BoundedDim d => Enum (Idx d)
deriving instance BoundedDim d => Integral (Idx d)
deriving instance BoundedDim d => Real (Idx d)
{- |
Although @Num (Idx d)@ requires @BoundedDim d@, it does not use @maxBound@
when @(d ~ XN m)@.
That is, if @(d ~ XN m)@ then @i = fromIntegral n@ always succeeds.
 -}
deriving instance BoundedDim d => Num (Idx d)

#else

{-# ANN xnatNInstEnumIdx (ToInstance NoOverlap) #-}
xnatNInstEnumIdx ::
       forall (n :: Nat)
     . KnownDim n => Dict (Enum (Idx (N n)))
xnatNInstEnumIdx = unsafeCoerce (Dict @(Enum (Idx n)))

-- ... other instances via constraints-deriving plugin
#endif
```

In practice, I think this shouldn't hold us back on getting rid of `HADDOCK_VERSION`, despite it being a useful case of HADDOCK_VERSION, since it is the last one in the whole of package.
Rather, `constraints-deriving` could add something like `ToInstanceWithDoc` that also takes the haddocks to attach to the generated instance as an argument.

michaelpj commented:
Personally, those seem like sufficiently minor uses that I think we could just say "sorry, this feature doesn't exist any more". Then there's a separate question of whether for example we would _want_ to add a feature like "hide overloaded label definitions".

alt-romes commented:
> Personally, those seem like sufficiently minor uses that I think we could just say "sorry, this feature doesn't exist any more". Then there's a separate question of whether for example we would _want_ to add a feature like "hide overloaded label definitions".

Great, I agree.
I'm going to proceed with getting rid of `HADDOCK_VERSION` in cabal altogether.

michaelpj commented:
Would probably be good to hear opinions from the Cabal maintainers though, I'm just being a busybody here.

alt-romes commented:
> Would probably be good to hear opinions from the Cabal maintainers though, I'm just being a busybody here.

Of course. @Kleidukos @gbaz @ffaf1 @Mikolaj I would appreciate your input.

I've opened [a haddock ticket](https://gitlab.haskell.org/ghc/ghc/-/issues/24819) to track/discuss a possible feature that allows hiding certain definitions by type.

Mikolaj commented:
IMHO, as long as we ensure a sufficient deprecation/overlap period so that maintainers of the other packages can adapt, let's move forward. 

How do we synchronism haddock and cabal releases to make sure the overlap of the added/removed features is sufficient? If haddock is not ready but we need to release a new cabal from master, what do we do? Revert a single PR on master or is this much harder?

alt-romes commented:
> IMHO, as long as we ensure a sufficient deprecation/overlap period so that maintainers of the other packages can adapt, let's move forward.

Providing a deprecation story here seems quite complicated. We can't easily tell whether a package makes use of `HADDOCK_VERSION` -- we blindly pass `-DHADDOCK_VERSION` to all haddock invocations.

Moreover, my opinion is that shipping a Cabal version with either new flags or hacks to provide both `HADDOCK_VERSION` to some packages *and* avoiding recompilation otherwise would be more costly than what it is worth. Even less so when you consider that afterwards you'd need yet another deprecation for the flag (like `haddock-recompilation`) we introduced to work around this. This whole cycle would churn devs for very little gain IMO!

> How do we synchronism haddock and cabal releases to make sure the overlap of the added/removed features is sufficient? If haddock is not ready but we need to release a new cabal from master, what do we do? Revert a single PR on master or is this much harder?

Cabal supports a window of haddock versions just like it supports a window of GHC versions. Nothing should change in that regard as the feature implemented in this PR is hidden behind a version guard. In particular, `HADDOCK_VERSION` has nothing to do with the haddock version being used since it is a CPP option created by cabal that haddock doesn't even know about.

---

I agree that we can't leave users without an option to upgrade while preserving the behaviour they currently have. Therefore, we need to ensure there is a reasonable *migration* story (over deprecation in this case, rather than alongside it).
I propose to implement myself a haddock feature to allow hiding of given definitions before merging this in.
This will allow haskell-gi to migrate to newer Cabal versions that don't pass `-DHADDOCK_VERSION` as long as a recent haddock is used. I believe `constraints-deriving` can easily extend their API to allow custom haddocks for the generate instances, which doesn't require changes to the compilation toolchain.

Mikolaj commented:
> Cabal supports a window of haddock versions just like it supports a window of GHC versions.

That's a good argument.

A pity we can't just issue a warning when `HADDOCK_VERSION` is used, but I can now see why this is impossible. So, instead, maybe we should issue a warning in 3.12.1.0 release notes that we intend to drop this flag in 3.14 and link to this ticket?

Thank you for looking into helping the actual identified user(s) of this feature. If anybody uses it in a closed code, I hope they can get enough guidance from this ticket and the solution for `haskell-gi`.

michaelpj commented:
> I agree that we can't leave users without an option to upgrade while preserving the behaviour they currently have. 

We could _ask_ the (very few) actual users and see if they care? If they don't care, then it would be silly to spend the time writing a migration path that nobody will use.

alt-romes commented:
I've extended this PR to include the two following changes:

* Cabal will no longer pass `-D__HADDOCK_VERSION__` to Haddock so as to avoid recompilation; this is a breaking change which we think is justifiable given there is a suitable migration strategy for all (very few) known use cases of this macro. I've edited the Cabal 3.12.1.0 release notes to announce this change will take place in the next release.

* Cabal will pass `--no-tmp-comp-dir` to haddock >= 2.18 by default to allow haddock to use the existing interface files produced by GHC.

I believe these changes address the main concerns with this PR, and put it in a mergeable state. The unaddressed concern is   the sanity check that haddock is not overwriting intermediate results produced by GHC, which I think is acceptable to be left unaddressed in this MR (I can open a ticket for it if you'd like me to).

Mikolaj commented on release-notes/WIP-Cabal-3.12.x.0.md:
```diff
@@ -0,0 +1,19 @@
+Pre-release Cabal 3.12.1.0 changelog and release notes.
```
Thanks a lot for the release note. Could you remove this first line? It made sense for the pre-release of cabal-install 3.12, but we are/were not pre-releasing Cabal the library.

alt-romes commented:
I've reached out to Iñaki by e-mail -- haskell-gi is supportive of this change forward, and is happy with the suggested migration story for omitting certain declarations given in https://gitlab.haskell.org/ghc/ghc/-/issues/24819; Iñaki has commented in that thread.

Moreover, I've addressed the concern of haddock overwriting GHC objects by first copying the output files into a temporary directory and running haddock there. A potential longer-term solution is for haddock to grow a flag that guarantees no files are overwritten regardless of configuration changes, but I'm not proposing to implement this myself in the immediate future -- the proposed solution can be made readily available now regardless of the haddock version.

I have to note that we can copy and run haddock on temporary directories o-files and hi-files, but not stub-files. The `-stubdir` flag has to be the same as the one used by GHC because `-stubdir` is (perhaps unfortunately) part of the recompilation fingerprint. Luckily, I don't see how stub files could create ABI-incompability problems, unlike object and interface files, so there's no problem if haddock (in calling GHC) were to overwrite them accidentally (e.g. if a haddock user-option triggered recompilation).

I believe this PR is ready to review and merge.

Kleidukos commented:
The changelog step needs to be rewritten to use the bindist of `changelog-d`. 
cc @fgaz

geekosaur commented on Cabal/src/Distribution/Simple/Haddock.hs:
```diff
@@ -800,6 +787,63 @@ getGhcLibDir verbosity lbi = do
     _ -> error "haddock only supports GHC and GHCJS"
   return $ mempty{argGhcLibDir = Flag l}
 
+-- | If hi-haddock is supported, this function creates temporary directories
```
```suggestion
-- | If Hi Haddock is supported, this function creates temporary directories
```
For consistency.

geekosaur commented on Cabal/src/Distribution/Simple/Haddock.hs:
```diff
@@ -800,6 +787,63 @@ getGhcLibDir verbosity lbi = do
+-- and copies existing interface and object files produced by GHC into them,
+-- then passes them off to the given continuation.
+--
+-- If hi-haddock is not supported, we can't re-use GHC's compilation files.
```
```suggestion
-- If Hi Haddock is _not_ supported, we can't re-use GHC's compilation files.
```

fgaz commented:
> The changelog step needs to be rewritten to use the bindist of `changelog-d`. cc @fgaz

oof. they were way too quick to update ghc. I'll have a look tomorrow.

Kleidukos commented:
@fgaz I think we can get away with it by just fetching the binary release.

alt-romes commented:
I've pushed a commit to this branch which builds `changelog-d` with `--allows-newer=base` rather than `--allows-newer=changelog-d:base` (ie it allows newer bases regardless of `< 4.20` bounds on base for the remaining packages in the plan, not just the bound on changelog-d).

This fixes the changelog job and unblocks CI.

Kleidukos commented on Cabal/src/Distribution/Simple/Haddock.hs:
```diff
@@ -800,6 +787,63 @@ getGhcLibDir verbosity lbi = do
+
+          act (tmpObjDir, tmpHiDir, fromFlag $ ghcOptHiDir vanillaOpts)
+
+  -- NB: Turns out, -stubdir is included in GHC's recompilation fingerprint.
+  -- This means that if we use a temporary directory for stubfiles
+  -- produced by GHC for the haddock invocation, haddock will trigger
+  -- full recompilation since the stubdir would be different.
+  --
+  -- So we don't use a temporary stubdir, despite the tmp o-dir and hi-dir:
+  --
+  -- We want to avoid at all costs haddock accidentally
+  -- overwriting o-files and hi-files (e.g. if a user specified
+  -- haddock-option triggers recompilation), and thus copy them to a
+  -- temporary directory to pass them on to haddock. However, stub files
+  -- are much less problematic since ABI-incompatibility isn't at play
+  -- here, that is, I don't think there exists a GHC flag that could
+  -- accidentally make a stub file incompatible with the one produced by
+  -- GHC from the same module.
```
This is an interesting piece of detail. Shouldn't it live in the `Hi Haddock Recompilation Avoidance` note?

alt-romes replied:
I'll do that.

Kleidukos commented on Cabal/src/Distribution/Simple/Haddock.hs:
```diff
@@ -800,6 +787,63 @@ getGhcLibDir verbosity lbi = do
+  -- haddock-option triggers recompilation), and thus copy them to a
+  -- temporary directory to pass them on to haddock. However, stub files
+  -- are much less problematic since ABI-incompatibility isn't at play
+  -- here, that is, I don't think there exists a GHC flag that could
```
Please favour the passive voice. :)  

```suggestion
  -- here, that is, there doesn't seem to be a GHC flag that could
```

Kleidukos commented on release-notes/WIP-Cabal-3.12.x.0.md:
Did I miss something these last releases? I thought we created the release notes from the changelog, with the significant changes having the:

```
significance: significant
```

metadata in their changelog entry.

alt-romes replied:
I did this in the style of `WIP-cabal-install-...`. This bit should stand out from the rest of the changelog regardless, and I think @Mikolaj suggested I do it like this.

Kleidukos replied:
Peculiar, I thought the `significance` metadata was enough. Thanks for the enlightenment!

alt-romes replied:
It may well be, but in this case I've done it this other way to be more particular about what is included separately in the release notes.

This should be the last outstanding point, let me know if we can mark this merge ready.

alt-romes replied:
I would appreciate it perhaps if you could run the manual QA notes actually. I've had success myself, but it'd be good to double check.

Mikolaj replied:
> I did this in the style of `WIP-cabal-install-...`. This bit should stand out from the rest of the changelog regardless, and I think @Mikolaj suggested I do it like this.

Now that I looked at my suggestion, the idea was to write it in a way that is not possible with changelogs, in order to turn the user attention to the impending breakage. Namely, changelogs focus on what's changed and we want to signal things that are going to change in the future (that would be a non-existent "Planned Significant Changes" section, not the usual "Significant Changes"). I guess just adding a single sentence close to the start of the release notes, pointing for details to the bigger blurb below, would accomplish the task.

Mikolaj replied:
> > I did this in the style of `WIP-cabal-install-...`. This bit should stand out from the rest of the changelog regardless, and I think @Mikolaj suggested I do it like this.
> 
> Now that I looked at my suggestion, the idea was to write it in a way that is not possible with changelogs, in order to turn the user attention to the impending breakage. Namely, changelogs focus on what's changed and we want to signal things that are going to change in the future (that would be a non-existent "Planned Significant Changes" section, not the usual "Significant Changes"). I guess just adding a single sentence close to the start of the release notes, pointing for details to the bigger blurb below, would accomplish the task.

Unless we are doing all the changes outright in 3.12, instead of deprecating now, to be completely removed in the future?

Kleidukos reviewed (commented):
Thanks a lot @alt-romes! I left a couple of comments. :)

Kleidukos reviewed (approved):
Alright, I don't see anything more to correct. I'll nonetheless call to @gbaz to take a last look, but at this stage I think we may be able to proceed forward to QA.

Kleidukos commented:
ping @haskell/manual-qa, time to shine!

jgotoh commented:
Manual QA was successful on WSL (Ubuntu 22.04, Windows 11).

Here is the log for anyone interested:

**Reproduce Bug with cabal latest (3.10.3.0)**

Resolving dependencies...
Build profile: -w ghc-9.8.2 -O1
In order, the following will be built (use -v for more details):
 - acme-missiles-0.3 (lib) (requires download & build)
 - deps-0.1.0 (lib) (first run)
Downloading  acme-missiles-0.3
Downloaded   acme-missiles-0.3
Starting     acme-missiles-0.3 (lib)
Building     acme-missiles-0.3 (lib)
Haddock      acme-missiles-0.3 (lib)
Installing   acme-missiles-0.3 (lib)
Completed    acme-missiles-0.3 (lib)
Configuring library for deps-0.1.0..
Preprocessing library for deps-0.1.0..
Building library for deps-0.1.0..
**[1 of 1] Compiling Lib**              ( Lib.hs, /home/.../dist-newstyle/build/x86_64-linux/ghc-9.8.2/deps-0.1.0/build/Lib.o, /home/.../dist-newstyle/build/x86_64-linux/ghc-9.8.2/deps-0.1.0/build/Lib.dyn_o )
Preprocessing library for deps-0.1.0..
Running Haddock on library for deps-0.1.0..
Old interface file was invalid:
    Exception when reading interface file  /tmp/.haddock-283725/Lib.hi
      /tmp/.haddock-283725/Lib.hi: withBinaryFile: does not exist (No such file or directory)
**[1 of 1] Compiling Lib**              ( Lib.hs, nothing )
Haddock coverage:
 100% (  3 /  3) in 'Lib'
Documentation created:
/home/julian/.../dist-newstyle/build/x86_64-linux/ghc-9.8.2/deps-0.1.0/doc/html/deps/

^ lib was compiled twice.

**Reproduce Fix with cabal head**

Resolving dependencies...
Build profile: -w ghc-9.8.2 -O1
In order, the following will be built (use -v for more details):
 - deps-0.1.0 (lib) (first run)
Warning: this is a debug build of cabal-install with assertions enabled.
Configuring library for deps-0.1.0...
Warning: this is a debug build of cabal-install with assertions enabled.
Preprocessing library for deps-0.1.0...
Building library for deps-0.1.0...
**[1 of 1] Compiling Lib**              ( Lib.hs, dist-newstyle/build/x86_64-linux/ghc-9.8.2/deps-0.1.0/build/Lib.o, dist-newstyle/build/x86_64-linux/ghc-9.8.2/deps-0.1.0/build/Lib.dyn_o )
Warning: this is a debug build of cabal-install with assertions enabled.
Preprocessing library for deps-0.1.0...
Running Haddock on library for deps-0.1.0...
Considering whether compilation is required for Lib:
Module flags unchanged
Optimisation flags changed; ignoring
HPC flags unchanged
signatures to merge in unchanged
[]
implementing module unchanged
Checking interface for module Acme.Missiles acme-missiles-0.3-ae693866e57964b306e10d08e46ec57afb464eb0ea11a3e9cd8b2d4d84af4b75
Module fingerprint unchanged
Checking interface for module Prelude base
Module fingerprint unchanged
Checking interface for module GHC.Types ghc-prim
Module fingerprint unchanged
Haddock coverage:
 100% (  3 /  3) in 'Lib'
Documentation created:
/home/..././dist-newstyle/build/x86_64-linux/ghc-9.8.2/deps-0.1.0/doc/html/deps/
Warning: this is a debug build of cabal-install with assertions enabled.

Lib.hs was compiled once ✅

jasagredo commented:
Unless I am doing something wrong, QA does not pass on Windows/MSYS2, it seems `Lib` is compiled twice:

```
➜ $(cygpath -m "C:\\Users\\Javier\\code\\cabal\\dist-newstyle\\build\\x86_64-windows\\ghc-9.8.2\\cabal-install-3.13.0.0\\x\\cabal\\build\\cabal\\cabal.exe") build
Warning: this is a debug build of cabal-install with assertions enabled.
Resolving dependencies...
Build profile: -w ghc-9.8.2 -O1
In order, the following will be built (use -v for more details):
 - deps-0.1.0 (lib) (first run)
Warning: this is a debug build of cabal-install with assertions enabled.
Configuring library for deps-0.1.0...
Warning: this is a debug build of cabal-install with assertions enabled.
Preprocessing library for deps-0.1.0...
Building library for deps-0.1.0...
[1 of 1] Compiling Lib              ( Lib.hs, dist-newstyle\build\x86_64-windows\ghc-9.8.2\deps-0.1.0\build\Lib.o )
Warning: this is a debug build of cabal-install with assertions enabled.
Preprocessing library for deps-0.1.0...
Running Haddock on library for deps-0.1.0...
Warning: The documentation for the following packages are not installed. No
links will be generated to these packages: array-0.5.6.0, base-4.19.1.0,
ghc-bignum-1.3, ghc-prim-0.11.0, stm-2.5.2.1
Considering whether compilation is required for Lib:
  Module flags have changed 723e9f5ce95d715375df6f4102da4501 -> 301ebdaa80354c8d4dc501aa439e15ea
[1 of 1] Compiling Lib              ( Lib.hs, nothing ) [Flags changed]
Haddock coverage:
 100% (  3 /  3) in 'Lib'
Warning: Lib: could not find link destinations for:
        - GHC.Types.IO
Documentation created:
C:\Users\Javier\code\bb\.\dist-newstyle\build\x86_64-windows\ghc-9.8.2\deps-0.1.0\doc\html\deps\
Warning: this is a debug build of cabal-install with assertions enabled.
```

jasagredo commented:
I don't see an obvious difference in the flags passed to GHC, more than the call to `ghc` compilation uses single backslashes and the `--optghc` args given to haddock have double backslashes:

```shell
❯ diff ghcargs optghc-haddockargs
4c4
< dist-newstyle\build\x86_64-windows\ghc-9.8.2\deps-0.1.0\build
---
> dist-newstyle\\build\\x86_64-windows\\ghc-9.8.2\\deps-0.1.0\\build
6c6
< dist-newstyle\build\x86_64-windows\ghc-9.8.2\deps-0.1.0\build
---
> dist-newstyle\\build\\x86_64-windows\\ghc-9.8.2\\deps-0.1.0\\haddock-objs-10336
8c8
< dist-newstyle\build\x86_64-windows\ghc-9.8.2\deps-0.1.0\build
---
> dist-newstyle\\build\\x86_64-windows\\ghc-9.8.2\\deps-0.1.0\\haddock-his-10336
10c10
< dist-newstyle\build\x86_64-windows\ghc-9.8.2\deps-0.1.0\build\extra-compilation-artifacts\hie
---
> dist-newstyle\\build\\x86_64-windows\\ghc-9.8.2\\deps-0.1.0\\build\\extra-compilation-artifacts\\hie
12c12
< dist-newstyle\build\x86_64-windows\ghc-9.8.2\deps-0.1.0\build
---
> dist-newstyle\\build\\x86_64-windows\\ghc-9.8.2\\deps-0.1.0\\build
15,21c15,21
< -idist-newstyle\build\x86_64-windows\ghc-9.8.2\deps-0.1.0\build
< -idist-newstyle\build\x86_64-windows\ghc-9.8.2\deps-0.1.0\build\autogen
< -idist-newstyle\build\x86_64-windows\ghc-9.8.2\deps-0.1.0\build\global-autogen
< -Idist-newstyle\build\x86_64-windows\ghc-9.8.2\deps-0.1.0\build\autogen
< -Idist-newstyle\build\x86_64-windows\ghc-9.8.2\deps-0.1.0\build\global-autogen
< -Idist-newstyle\build\x86_64-windows\ghc-9.8.2\deps-0.1.0\build
< -IC:\msys64\ucrt64\include
---
> -idist-newstyle\\build\\x86_64-windows\\ghc-9.8.2\\deps-0.1.0\\build
> -idist-newstyle\\build\\x86_64-windows\\ghc-9.8.2\\deps-0.1.0\\build\\autogen
> -idist-newstyle\\build\\x86_64-windows\\ghc-9.8.2\\deps-0.1.0\\build\\global-autogen
> -Idist-newstyle\\build\\x86_64-windows\\ghc-9.8.2\\deps-0.1.0\\build\\autogen
> -Idist-newstyle\\build\\x86_64-windows\\ghc-9.8.2\\deps-0.1.0\\build\\global-autogen
> -Idist-newstyle\\build\\x86_64-windows\\ghc-9.8.2\\deps-0.1.0\\build
> -IC:\\msys64\\ucrt64\\include
23c23
< -optPdist-newstyle\build\x86_64-windows\ghc-9.8.2\deps-0.1.0\build\autogen\cabal_macros.h
---
> -optPdist-newstyle\\build\\x86_64-windows\\ghc-9.8.2\\deps-0.1.0\\build\\autogen\\cabal_macros.h
30c30
< C:\Users\Javier\AppData\Local\cabal\store\ghc-9.8.2-7294\package.db
---
> C:\\Users\\Javier\\AppData\\Local\\cabal\\store\\ghc-9.8.2-7294\\package.db
32c32
< C:\Users\Javier\code\bb\dist-newstyle\packagedb\ghc-9.8.2
---
> C:\\Users\\Javier\\code\\bb\\dist-newstyle\\packagedb\\ghc-9.8.2
34c34
< C:\Users\Javier\code\bb\.\dist-newstyle\build\x86_64-windows\ghc-9.8.2\deps-0.1.0\package.conf.inplace
---
> C:\\Users\\Javier\\code\\bb\\.\\dist-newstyle\\build\\x86_64-windows\\ghc-9.8.2\\deps-0.1.0\\package.conf.inplace
```

However computing the flag fingerprint uses `normalise` on those paths, so it should be fine?

Mikolaj commented:
Regarding a backport to 3.12, this PR seems to break the 3.12 Cabal the library API, so it's seems too late. I'd vote for 3.14.

jasagredo commented:
It works with a locally patched GHC 9.11 (HEAD as of today):
```
➜ $CABAL build --with-compiler /c/Users/Javier/code/ghc/_build/stage0/bin/ghc                
Warning: this is a debug build of cabal-install with assertions enabled.
Warning: Parsing the index cache failed (Data.Binary.Get.runGet at position
16: Non-matching structured hashes: f46da61e7afa58a5e8fd1d2b6fb79899;
expected: d81bdd513f41b5d7ee4cd28455adadbe). Trying to regenerate the index
cache...
Resolving dependencies...
Build profile: -w ghc-9.11.20240607 -O1
In order, the following will be built (use -v for more details):
 - acme-missiles-0.3 (lib) (requires build)
 - deps-0.1.0 (lib) (first run)
Starting     acme-missiles-0.3 (lib)
Building     acme-missiles-0.3 (lib)
Haddock      acme-missiles-0.3 (lib)
Installing   acme-missiles-0.3 (lib)
Completed    acme-missiles-0.3 (lib)
Warning: this is a debug build of cabal-install with assertions enabled.
Configuring library for deps-0.1.0...
Warning: this is a debug build of cabal-install with assertions enabled.
Preprocessing library for deps-0.1.0...
Building library for deps-0.1.0...
[1 of 1] Compiling Lib              ( Lib.hs, dist-newstyle\build\x86_64-windows\ghc-9.11.20240607\deps-0.1.0\build\Lib.o )
MOD FLAGS: mainis:          Nothing
MOD FLAGS: safeHS:          TrustInfo None
MOD FLAGS: lang:            (Just 0,[5,7,23,44,98,99,101,120,122,125,130])
MOD FLAGS: cpp includes:
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240607\deps-0.1.0\build
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240607\deps-0.1.0\build
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240607\deps-0.1.0\build\autogen
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240607\deps-0.1.0\build\global-autogen
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240607\deps-0.1.0\build
  C:\msys64\ucrt64\include
MOD FLAGS: pic_p_opts:      []
MOD FLAGS: optP sig:        ([],19b7ece0dd17ba28e7ce435ed74624da)
MOD FLAGS: paths:           ["hc"]
MOD FLAGS: prof:            0
MOD FLAGS: ticky:           [False,False,False,False,False]
MOD FLAGS: debugLevel:      0
MOD FLAGS: callerCc empty:  True
MOD FLAGS: optP flags:      ["dist-newstyle\\build\\x86_64-windows\\ghc-9.11.20240607\\deps-0.1.0\\build\\autogen\\cabal_macros.h","-include"]
OPT FLAGS: opt_flags:
  Opt_CallArity
  Opt_Strictness
  Opt_FullLaziness
  Opt_FloatIn
  Opt_Specialise
  Opt_CrossModuleSpecialise
  Opt_CSE
  Opt_StgCSE
  Opt_DoLambdaEtaExpansion
  Opt_IgnoreAsserts
  Opt_DoEtaReduction
  Opt_CaseMerge
  Opt_CaseFolding
  Opt_UnboxSmallStrictFields
  Opt_EnableRewriteRules
  Opt_CmmSink
  Opt_CmmElimCommonBlocks
  Opt_Loopification
  Opt_CfgBlocklayout
  Opt_CprAnal
  Opt_WorkerWrapper
  Opt_SolveConstantDicts
Warning: this is a debug build of cabal-install with assertions enabled.
Preprocessing library for deps-0.1.0...
Running Haddock on library for deps-0.1.0...
Warning: The documentation for the following packages are not installed. No
links will be generated to these packages: array-0.5.6.0, base-4.20.0.0,
ghc-bignum-1.3, ghc-internal-9.1001.0, ghc-prim-0.11.0, stm-2.5.3.0
Considering whether compilation is required for Lib:
MOD FLAGS: mainis:          Nothing
MOD FLAGS: safeHS:          TrustInfo None
MOD FLAGS: lang:            (Just 0,[5,7,23,44,98,99,101,120,122,125,130])
MOD FLAGS: cpp includes:
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240607\deps-0.1.0\build
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240607\deps-0.1.0\build
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240607\deps-0.1.0\build\autogen
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240607\deps-0.1.0\build\global-autogen
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240607\deps-0.1.0\build
  C:\msys64\ucrt64\include
MOD FLAGS: pic_p_opts:      []
MOD FLAGS: optP sig:        ([],19b7ece0dd17ba28e7ce435ed74624da)
MOD FLAGS: paths:           ["hc"]
MOD FLAGS: prof:            0
MOD FLAGS: ticky:           [False,False,False,False,False]
MOD FLAGS: debugLevel:      0
MOD FLAGS: callerCc empty:  True
MOD FLAGS: optP flags:      ["dist-newstyle\\build\\x86_64-windows\\ghc-9.11.20240607\\deps-0.1.0\\build\\autogen\\cabal_macros.h","-include"]
Module flags unchanged
OPT FLAGS: opt_flags:
  Opt_DoLambdaEtaExpansion
  Opt_DoEtaReduction
Optimisation flags changed; ignoring
HPC flags unchanged
signatures to merge in unchanged
[]
implementing module unchanged
Checking interface for module Acme.Missiles acme-missiles-0.3-9d2bfaea19d60d250e0d2dd16d3ccb0608c816c1
Module fingerprint unchanged
Checking interface for module Prelude base
Module fingerprint unchanged
Checking interface for module GHC.Types ghc-prim
Module fingerprint unchanged
Haddock coverage:
 100% (  3 /  3) in 'Lib'
Warning: Lib: could not find link destinations for: 
        - GHC.Types.IO
Documentation created:
C:\Users\Javier\code\bb\.\dist-newstyle\build\x86_64-windows\ghc-9.11.20240607\deps-0.1.0\doc\html\deps\
Warning: this is a debug build of cabal-install with assertions enabled.
```

Mikolaj commented:
@jasagredo: so it's good news that it works, but bad news that it didn't help us pinpoint the problem when run with GHC 9.8.2? How could, e.g., @FinleyMcIlwaine or @alt-romes help you diagnose further?

jasagredo commented:
I will try tomorrow with ghc 9.8. @FinleyMcIlwaine provided me with the tracing patch today at Zurihac.

Mikolaj commented:
Might this be related by chance? https://gitlab.haskell.org/ghc/ghc/-/issues/24964

jasagredo commented:
The library had no foreign symbols 🤔. It turns out to be even more puzzling than anticipated.

So a locally built GHC 9.8.2 does not show the bug. The bindist 9.8.2 and 9.6.5 GHCs do show the bug, which makes no sense, even more because in Windows it will use an in-tree clang and libs, so it shouldn't depend on the environment in which it is being built.

I looked at this with Ben too, with no luck. His suggestion was to build GHC again tomorrow with `--flavour=release` which would be the same as the distributed bindist, but if that shows a difference then it will probably be worrying 😄

jasagredo commented:
GHC 9.8.2 built with `hadrian/build.bat --flavour=release` says "Module flags unchanged". Now compiling GHC 9.8.2 with `hadrian/build.bat --flavour=release --hash-unit-ids` and praying :pray:

Mikolaj commented:
@jasagredo: Was the answer to the prayers to your liking?

michaelpj commented:
@alt-romes and I found some sad issues with this at Zurihac. Notably:

Turning on `documentation: true` will still re-build all your dependencies, despite the previously built interface files containing all the needed information to produce the Haddock. This rebuild will _also_ force a rebuild of your local modules, since the hashes of the dependencies will have changed!

So in practice this only benefits you for changes to your _local_ files, which is a shame. It seems like it gives us the tools to avoid rebuilding dependencies (and your local modules!) when turning on documentation, but we don't currently do things in the right way for that to happen.

(Probably this should be in a new issue...)

phadej commented:
`documentation: true` issue is discussed in https://github.com/haskell/cabal/issues/8725

While good intention, in practice it doesn't work, and you cannot even override it.

> What is worse, if someone doesn't want to enable documentation: True, there seems to be no way to permanently disable this behavior.

phadej commented:
Duncan alludes:

> Perhaps a nice solution (though not an easy solution to implement) would be if we could make the haddocks independently installable: so that building the docs for a dependency did not imply rebuilding the code. For example, similar to how we handle building profiling these days. 

With haddocks built from ordinary interface files, that should be more easily doable (but still not easy, as it's new codepaths)

michaelpj commented:
Yes, Rodrigo and I discussed that as a possibility, I think that would be the most graceful solution.

jasagredo commented:
@Mikolaj: 
> @jasagredo: Was the answer to the prayers to your liking?

It was not unfortunately. So any GHC I build locally, even using the same config options as the CI, says "Module flags unchanged".

Every GHC distributed *and* a GHC I downloaded from a very recent pipeline on GHC's gitlab says "Module flags changed".

I think I will try to apply @FinleyMcIlwaine's patch to GHC-head and get a binary with it from CI, as it seems to be the only building environment that can reproduce this behavior.

Such a mysterious issue.

michaelpj commented:
I made a new issue for the recompilation issue, since I think it's distinct from the question of whether `documentation: True` is enabled by default: https://github.com/haskell/cabal/issues/10111

jasagredo commented:
@FinleyMcIlwaine I got a CI GHC with your patch, and bingo!

```
➜ $CABAL build
Warning: this is a debug build of cabal-install with assertions enabled.
Resolving dependencies...
Build profile: -w ghc-9.11.20240614 -O1
In order, the following will be built (use -v for more details):
 - acme-missiles-0.3 (lib) (requires build)
 - deps-0.1.0 (lib) (first run)
Starting     acme-missiles-0.3 (lib)
Building     acme-missiles-0.3 (lib)
Haddock      acme-missiles-0.3 (lib)
Installing   acme-missiles-0.3 (lib)
Completed    acme-missiles-0.3 (lib)
Warning: this is a debug build of cabal-install with assertions enabled.
Configuring library for deps-0.1.0...
Warning: this is a debug build of cabal-install with assertions enabled.
Preprocessing library for deps-0.1.0...
Building library for deps-0.1.0...
[1 of 1] Compiling Lib              ( Lib.hs, dist-newstyle\build\x86_64-windows\ghc-9.11.20240614\deps-0.1.0\build\Lib.o )
MOD FLAGS: mainis:          Nothing
MOD FLAGS: safeHS:          TrustInfo None
MOD FLAGS: lang:            (Just 0,[5,7,23,44,98,99,101,120,122,125,130])
MOD FLAGS: cpp includes:
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240614\deps-0.1.0\build
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240614\deps-0.1.0\build
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240614\deps-0.1.0\build\autogen
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240614\deps-0.1.0\build\global-autogen
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240614\deps-0.1.0\build
  C:\msys64\ucrt64\include
MOD FLAGS: pic_p_opts:      []
MOD FLAGS: optP sig:        ([],8383540998520b6962a51feacf466dc3)
MOD FLAGS: paths:           ["hc"]
MOD FLAGS: prof:            0
MOD FLAGS: ticky:           [False,False,False,False,False]
MOD FLAGS: debugLevel:      0
MOD FLAGS: callerCc empty:  True
MOD FLAGS: optP flags:      ["dist-newstyle\\\\build\\\\x86_64-windows\\\\ghc-9.11.20240614\\\\deps-0.1.0\\\\build\\\\autogen\\\\cabal_macros.h","-include"]
OPT FLAGS: opt_flags:
  Opt_CallArity
  Opt_Strictness
  Opt_FullLaziness
  Opt_FloatIn
  Opt_Specialise
  Opt_CrossModuleSpecialise
  Opt_CSE
  Opt_StgCSE
  Opt_DoLambdaEtaExpansion
  Opt_IgnoreAsserts
  Opt_DoEtaReduction
  Opt_CaseMerge
  Opt_CaseFolding
  Opt_UnboxSmallStrictFields
  Opt_EnableRewriteRules
  Opt_CmmSink
  Opt_CmmElimCommonBlocks
  Opt_Loopification
  Opt_CfgBlocklayout
  Opt_CprAnal
  Opt_WorkerWrapper
  Opt_SolveConstantDicts
Warning: this is a debug build of cabal-install with assertions enabled.
Preprocessing library for deps-0.1.0...
Running Haddock on library for deps-0.1.0...
Considering whether compilation is required for Lib:
MOD FLAGS: mainis:          Nothing
MOD FLAGS: safeHS:          TrustInfo None
MOD FLAGS: lang:            (Just 0,[5,7,23,44,98,99,101,120,122,125,130])
MOD FLAGS: cpp includes:
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240614\deps-0.1.0\build
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240614\deps-0.1.0\build
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240614\deps-0.1.0\build\autogen
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240614\deps-0.1.0\build\global-autogen
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240614\deps-0.1.0\build
  C:\msys64\ucrt64\include
MOD FLAGS: pic_p_opts:      []
MOD FLAGS: optP sig:        ([],4fccfb1d242a0524c052c01a0ed15cdf)
MOD FLAGS: paths:           ["hc"]
MOD FLAGS: prof:            0
MOD FLAGS: ticky:           [False,False,False,False,False]
MOD FLAGS: debugLevel:      0
MOD FLAGS: callerCc empty:  True
MOD FLAGS: optP flags:      ["dist-newstyle\\build\\x86_64-windows\\ghc-9.11.20240614\\deps-0.1.0\\build\\autogen\\cabal_macros.h","-include"]
  Module flags have changed 0e22040b910adf908347fb5cb2731333 -> e5930a88675b596d7b784723a39a2246
[1 of 1] Compiling Lib              ( Lib.hs, nothing ) [Flags changed]
MOD FLAGS: mainis:          Nothing
MOD FLAGS: safeHS:          TrustInfo None
MOD FLAGS: lang:            (Just 0,[5,7,23,44,98,99,101,120,122,125,130])
MOD FLAGS: cpp includes:
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240614\deps-0.1.0\build
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240614\deps-0.1.0\build
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240614\deps-0.1.0\build\autogen
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240614\deps-0.1.0\build\global-autogen
  dist-newstyle\build\x86_64-windows\ghc-9.11.20240614\deps-0.1.0\build
  C:\msys64\ucrt64\include
MOD FLAGS: pic_p_opts:      []
MOD FLAGS: optP sig:        ([],4fccfb1d242a0524c052c01a0ed15cdf)
MOD FLAGS: paths:           ["hc"]
MOD FLAGS: prof:            0
MOD FLAGS: ticky:           [False,False,False,False,False]
MOD FLAGS: debugLevel:      0
MOD FLAGS: callerCc empty:  True
MOD FLAGS: optP flags:      ["dist-newstyle\\build\\x86_64-windows\\ghc-9.11.20240614\\deps-0.1.0\\build\\autogen\\cabal_macros.h","-include"]
OPT FLAGS: opt_flags:
  Opt_DoLambdaEtaExpansion
  Opt_DoEtaReduction
Haddock coverage:
 100% (  3 /  3) in 'Lib'
Documentation created:
C:\Users\Javier\code\bb\.\dist-newstyle\build\x86_64-windows\ghc-9.11.20240614\deps-0.1.0\doc\html\deps\
Warning: this is a debug build of cabal-install with assertions enabled.
```

So the diff:
```
➜ diff first second
12c12
< MOD FLAGS: optP sig:        ([],8383540998520b6962a51feacf466dc3)
---
> MOD FLAGS: optP sig:        ([],4fccfb1d242a0524c052c01a0ed15cdf)
18c18
< MOD FLAGS: optP flags:      ["dist-newstyle\\\\build\\\\x86_64-windows\\\\ghc-9.11.20240614\\\\deps-0.1.0\\\\build\\\\autogen\\\\cabal_macros.h","-include"]
---
> MOD FLAGS: optP flags:      ["dist-newstyle\\build\\x86_64-windows\\ghc-9.11.20240614\\deps-0.1.0\\build\\autogen\\cabal_macros.h","-include"]
```

jasagredo commented:
Upstream issue https://gitlab.haskell.org/ghc/ghc/-/issues/24988

LeventErkok commented:
@FinleyMcIlwaine I do want to have access to `__HADDOCK_VERSION`; alas, using the mechanism described here doesn't seem to work.

To be specific, I created a `project.cabal` file, with contents:

```
packages: .

package *
  documentation: True
  haddock-version-cpp: true
```

but this doesn't seem to make `__HADDOCK_VERSION` available in source files. Any advice?
